<a href="https://colab.research.google.com/github/AntonDozhdikov/MFTI/blob/main/%5Bseminar%2Cbasic%5Dbash.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

In [3]:
!apt-get moo

                        (__) 
                        (oo) 
                  /------\/ 
                 / |    ||   
                *  /\---/\ 
                   ~~   ~~   
..."It's a Bird ... It's a Plane ... It's Super Cow!"...


В этом ноутбуке мы рассмотрим базовые команды командной строки в Linux

План
  - Команды UNIX
  - Файловая система
  - Аргументы к командам
  - Переменные окружения


# Команды в UNIX

Линукс принадлежит большому семейству операционных систем под названием UNIX, коорые происходят от ОС AT&T Unix, разроботанной еще в 1970х. Mac OS, кстати тоже ее потомок.

В операционных системах UNIX-типа для взаимодействия с ОС используют специальные команды. До появления окошек и кнопочек люди создавали файлы и запускали команды исключительно с помощью этих команд. 

Сейчас этот способ тоже исользуют, тк есть ряд преимуществ:
 - Намного более удобное подключение к компьютеру по сети
 - Иногда удобнее запустить одну команду, чем бродить по менюшкам в поисках нужной опции.
 - Ломанды можно составлять в скрипты и автоматизировать задачи
 



Небольшой ликбез по понятиям:

- **Командная строка** - вид интерфейса, когда вы набираете текстовые команды
- **Терминал** - Программа, которая отображает строку. Например: *gnome-terminal*, *xterm*, *tilda*
- **Оболочка (shell)** - Программа которая выполняет команды. Например: *bash*, *zsh*, *ssh*
- **Консоль** - Железо, например клавиатура и монитор.

Запускать команды можно как из терминала, так и из юпитера или колаба, поставив перед командой !

In [4]:
!echo "Hello World"

Hello World


# Файловая система

В Unix системах все файлы образуют структуру похожую на дерево, в каждой директории(папке) хранятся файлы и другие директории. Самая первая директория которая содержит все остальные, называется корневой (*root*)


У каждой директории есть путь (по сути, просто строка) который показывает где она находится в этом дереве. "/" - это путь корневой директории. 

---



![alt text](http://linuxcommand.org/images/file_manager.jpg)

В командной строке все команды выполняются в какой-либо директории.

Представьте, что дерево файловой системы - это лабиринт, и каждая директория - это комната с файлами и дверьми в другие директории. В каждый момент времени вы находитесь в одной комнате.

Посмотрм, как нам ходить по нашему лабиринту.

In [5]:
# Выведем путь текущей директории
!pwd

/content


In [6]:
# Посмотрим что в нашей директории есть
!ls 
#ls is for "list"

sample_data


Некоторые директории хранят файлы необходимые для работы программ. В них нужно что-либо менять только если значешь, что делаешь. 

Это так называемые скрытые директории (hidden или dot directories/files).

In [7]:
#

In [8]:
#Создадим пустой файлик
! touch test-file
# пробел в имени - плохая идея, создадутся два файла
! touch hello world

# посмотрим что получилось
! ls

hello  sample_data  test-file  world


In [9]:
#Удаление файлов
! rm world
! ls

hello  sample_data  test-file


In [10]:
# Новая директория
! mkdir please_come_here


Некоторые директории хранят файлы необходимые для работы программ. В них нужно что-либо менять только если значешь, что делаешь.

Это так называемые скрытые директории (hidden или dot directories/files).


In [11]:
# Покажем больше информации и отобразим скрытые файлы
!ls -al
# Первое слово - команда, а второе - аргумент

total 20
drwxr-xr-x 1 root root 4096 Feb 18 06:18 .
drwxr-xr-x 1 root root 4096 Feb 18 06:15 ..
drwxr-xr-x 4 root root 4096 Feb 17 19:29 .config
-rw-r--r-- 1 root root    0 Feb 18 06:17 hello
drwxr-xr-x 2 root root 4096 Feb 18 06:18 please_come_here
drwxr-xr-x 1 root root 4096 Feb 17 19:30 sample_data
-rw-r--r-- 1 root root    0 Feb 18 06:17 test-file




```
.  - это путь к той директории в которой мы уже находимся
.. - это путь к родительской директории
.config - это скрытая папка
```

Каждая строчка - это одна из "дверей" в другую директорию либо файл.
*Относительный* путь это не полный "адрес" директории а то, куда нам нужно пойти из места в котором мы сейчас находимся. 


```
 drwxr-xr-x 1 root root 4096 Feb 15 17:21 please_come_here
 |                           дата создания    |
 |                                         _относительный_ путь
 |
  \ эта метка значит, что это путь к другой директории
```



In [12]:
# переход в другую директорию команда cd
! pwd
! cd please_come_here
! pwd

/content
/content


Заработало?

Не должно. Тут проблема в том что юпитер запускает каждый раз новую оболочку в своей директории (той, откуда его вызвали) , и после завершения команды она закрывается.

Чтоб команды запускались в новой директории нужно поменять директорию самого юпитера.


```
%cd
```



In [ ]:
! pwd
% cd please_come_here
! pwd

Когда мы переходим в другую директорию, ее имя добавляется к нашему пути через "/"

Например, были в `/home`, перешли в директорию `Pictures `оказались в `/home/Pictures`

Абсолютный путь файла в этой директории будет /home/Picutes/image.png 

Если мы хотим взять файл из директории, которая является дочерней к той директории, где мы сейчас находимся, мы можем указать относительный путь файла.

Например, из директории `/home` мы должны обращаться к нему по пути `Pictures/image.png`

In [ ]:
# команды можно комбинировать в одну символом &
! ls & ls -a & pwd
%cd ..

# Чтоб команды выполнялись одна за одной нужно использовать &&
! pwd && cd please_come_here && pwd

## Запись и чтение в файлы

In [13]:
# Вывести текст в консоль

!echo "Sample text Пример текста"
# Echo - потому что команда выводит на выход то же, что и получает

Sample text Пример текста


In [14]:
# Вместо консоли отправить в файл

! echo "London is the capital of Great Britain" >> English.txt


In [15]:
# Прочитать файл и вывести в консоль

! cat English.txt

London is the capital of Great Britain


## Скачивание из интернета

In [22]:
# сохраняем старницу по адресу

! curl example.com >> example-com.html
! ls
! cat example-com.html


  % Total    % Received % Xferd  Average Speed   Time    Time     Time  Current
                                 Dload  Upload   Total   Spent    Left  Speed
100  1256  100  1256    0     0  22836      0 --:--:-- --:--:-- --:--:-- 22836
English.txt	  hello       please_come_here	test-file
example-com.html  insta.html  sample_data
<!doctype html>
<html>
<head>
    <title>Example Domain</title>

    <meta charset="utf-8" />
    <meta http-equiv="Content-type" content="text/html; charset=utf-8" />
    <meta name="viewport" content="width=device-width, initial-scale=1" />
    <style type="text/css">
    body {
        background-color: #f0f0f2;
        margin: 0;
        padding: 0;
        font-family: -apple-system, system-ui, BlinkMacSystemFont, "Segoe UI", "Open Sans", "Helvetica Neue", Helvetica, Arial, sans-serif;
        
    }
    div {
        width: 600px;
        margin: 5em auto;
        padding: 2em;
        background-color: #fdfdff;
        border-radius: 0.5em;
        box-

In [20]:
! curl instagram.com

Иногда сервера пытаются перенаправить запрос на другой адрес. Curl не следует по таким перенаправлениям. Есть прогаммка wget которая может переходить по таким ссылкам и вообще проходить по всем ссылкам на странице и скачивать их.

In [21]:
! wget instagram.com -O insta.html

URL transformed to HTTPS due to an HSTS policy
--2023-02-18 06:19:39--  https://instagram.com/
Resolving instagram.com (instagram.com)... 157.240.254.174, 2a03:2880:f275:1e9:face:b00c:0:4420
Connecting to instagram.com (instagram.com)|157.240.254.174|:443... connected.
HTTP request sent, awaiting response... 301 Moved Permanently
Location: https://www.instagram.com/ [following]
--2023-02-18 06:19:39--  https://www.instagram.com/
Resolving www.instagram.com (www.instagram.com)... 157.240.254.174, 2a03:2880:f275:e9:face:b00c:0:4420
Connecting to www.instagram.com (www.instagram.com)|157.240.254.174|:443... connected.
HTTP request sent, awaiting response... 302 Found
Location: https://www.facebook.com/unsupportedbrowser [following]
--2023-02-18 06:19:39--  https://www.facebook.com/unsupportedbrowser
Resolving www.facebook.com (www.facebook.com)... 157.240.254.35, 2a03:2880:f175:181:face:b00c:0:25de
Connecting to www.facebook.com (www.facebook.com)|157.240.254.35|:443... connected.
HTTP re

Обратите внимание на

`HTTP request sent, awaiting response... 301 Moved Permanently`

# Аргументы и параметры к командам

Обычно аргументы – это то, над чем выполняется команда. Например, команда перехода в другую директорию должна получить в качестве аргумента путь директории, куда следует перейти. Команде вывода на экран содержимого файла надо передать имя файла.

Параметры модифицируют, как бы видоизменяют работу программы. Например, с помощью команды просмотра содержимого каталога можно вывести простой список вложенных объектов. Используя же определенный параметр, можно получить подробные сведения о каждом файле и папке.

Аргументы идут через пробел как отдельные слова

Параметры идут после дефиса: ` -f` (однобуквенные) либо после двух дефисов: `--file`  

In [24]:
# календарик на весь год

!ncal 2023


                                  2023
    January           February          March             April             
Su  1  8 15 22 29        5 12 19 26        5 12 19 26        2  9 16 23 30
Mo  2  9 16 23 30        6 13 20 27        6 13 20 27        3 10 17 24   
Tu  3 10 17 24 31        7 14 21 28        7 14 21 28        4 11 18 25   
We  4 11 18 25        1  8 15 22        1  8 15 22 29        5 12 19 26   
Th  5 12 19 26        2  9 16 23        2  9 16 23 30        6 13 20 27   
Fr  6 13 20 27        3 10 17 24        3 10 17 24 31        7 14 21 28   
Sa  7 14 21 28        4 11 18 25        4 11 18 25        1  8 15 22 29   

    May               June              July              August            
Su     7 14 21 28        4 11 18 25        2  9 16 23 30     6 13 20 27   
Mo  1  8 15 22 29        5 12 19 26        3 10 17 24 31     7 14 21 28   
Tu  2  9 16 23 30        6 13 20 27        4 11 18 25     1  8 15 22 29   
We  3 10 17 24 31        7 14 21 28        5 12 19 26   

In [25]:
# передаем 2 аргумента

!ncal 05 2023


    May 2023          
Su     7 14 21 28   
Mo  1  8 15 22 29   
Tu  2  9 16 23 30   
We  3 10 17 24 31   
Th  4 11 18 25      
Fr  5 12 19 26      
Sa  6 13 20 27      


In [26]:
# передаем параметр -b

!ncal -b 03 2023


     March 2023       
Su Mo Tu We Th Fr Sa  
          1  2  3  4  
 5  6  7  8  9 10 11  
12 13 14 15 16 17 18  
19 20 21 22 23 24 25  
26 27 28 29 30 31     
                      


In [27]:
# Однобуквенные параметры можно склеивать

!ncal -b -w 03 2019
!ncal -bw 03 2019

       March 2019          
 w| Su Mo Tu We Th Fr Sa   
 8|                 1  2   
 9|  3  4  5  6  7  8  9   
10| 10 11 12 13 14 15 16   
11| 17 18 19 20 21 22 23   
12| 24 25 26 27 28 29 30   
13| 31                     
       March 2019          
 w| Su Mo Tu We Th Fr Sa   
 8|                 1  2   
 9|  3  4  5  6  7  8  9   
10| 10 11 12 13 14 15 16   
11| 17 18 19 20 21 22 23   
12| 24 25 26 27 28 29 30   
13| 31                     


# Переменные окружения

Оболочка в себе содержит так называемые переменные окружения - глобальные константы которые доступны всем программам которые она выполняет. В них удобно хранить например путь к конфигурационным файлам или порт какого нибудь сервиса.

Переменные окружения зачастую определяют как ведет себя программа. Например, питон ищет пакеты в переменной `PYTHONPATH`

В переменной `PATH` находятся пути в которых Bash ищет исполняемые файлы для других программ


In [29]:
# обьявляем переменную и считываем ее ($ перед именем)
!FOO=kindersurprise && echo $FOO

kindersurprise


In [30]:
!echo $PATH
!echo $PYTHONPATH

/opt/bin:/usr/local/nvidia/bin:/usr/local/cuda/bin:/usr/local/sbin:/usr/local/bin:/usr/sbin:/usr/bin:/sbin:/bin:/tools/node/bin:/tools/google-cloud-sdk/bin
/env/python


Мы можем создать новую папку и заставить питон думать что в ней лежит модуль

In [31]:
! rm -r fakemodule
!mkdir fakemodule
!mkdir fakemodule/power
! echo "print('A great power comes with a great responsibility')" >> fakemodule/power/__init__.py




rm: cannot remove 'fakemodule': No such file or directory


In [ ]:
!export a="Эта переменная будет доступна другим программам вызванным в той же оболочке" && echo $a
# Но тут уже не получится, для каждой команды Jupyter запускает свою оболочку 
! echo $a

In [ ]:
! PYTHONPATH=$PYTHONPATH:$(pwd)/fakemodule && python -c 'import power'

конструкция `$(pwd)` подставляет вывод команды `pwd` в выражение. 


```
PYTHONPATH=$PYTHONPATH:$(pwd)/fakemodule

```
После подстановки переменных мы получим 

```
PYTHONPATH=<то что было в PYTHONPATH>:<текущая директория>/fakemodule

```
Это добавит нашу fakemodule в переменную окружения и позволит питону найти наш модуль

